File: Lab/SEM4/MPP/lab5/moto_java/src/main/java/org/zdroba/service/RaceEventServiceImpl.java
package org.zdroba.service;

import org.zdroba.entity.RaceEvent;
import org.zdroba.exceptions.AlreadyExistsException;
import org.zdroba.repository.RaceEventRepository;
import org.zdroba.repository.RaceEventRepositoryImpl;

import java.util.List;

public class RaceEventServiceImpl implements RaceEventService{

    private final RaceEventRepository raceEventRepository;

    public RaceEventServiceImpl(RaceEventRepository raceEventRepository) {
        this.raceEventRepository = raceEventRepository;
    }

    @Override
    public List<RaceEvent> getAll() {
        return raceEventRepository.getAll();
    }

    @Override
    public void add(int engine) throws AlreadyExistsException {
        RaceEvent raceEvent = raceEventRepository.find(engine);

        if(raceEvent!=null)
            throw new AlreadyExistsException(String.format("An event with engine: %d already exists", engine));

        raceEvent = new RaceEvent(engine);
        raceEventRepository.add(raceEvent); // object gets id returned in .add method
    }
}

File: Lab/SEM4/MPP/lab9/moto_c/src/Client/service/RaceEventServiceImpl.cs
using System.Collections.Generic;
using moto_c.Common.sync;
using moto_c.entity;
using moto_c.exceptions;
using moto_c.repository;
using moto_c.sync;

namespace moto_c.service;

public class RaceEventServiceImpl: RaceEventService
{
    private RaceEventRepository raceEventRepository;
    private readonly SocketNotifier notifier;

    public RaceEventServiceImpl(RaceEventRepository raceEventRepository, SocketNotifier notifier)
    {
        this.raceEventRepository = raceEventRepository;
        this.notifier = notifier;
    }

    public List<RaceEvent> getAll()
    {
        return raceEventRepository.getAll();
    }

    public void add(int engine)
    {
        RaceEvent? raceEvent = raceEventRepository.find(engine);

        if (raceEvent != null)
            throw new AlreadyExistsException("Event with this engine exists");

        raceEvent = new RaceEvent(engine);
        raceEventRepository.add(raceEvent);

        Request req = new Request(RequestType.EVENT_ADD, "{\"id\":" + raceEvent.id + "}");
        notifier.NotifyPeer(req.ToString());
    }
}